### DISCUSSION TITLE
How does carbonapi translate the grafana query to return data

### DISCUSSION THREAD

**Discussion by rickyari** [go-graphite/carbonapi#633] [Q&A]
We have a series which  creates multiple directories (over 5000) and the grafana query is used to visualize metrics. Just wanted to know how carbonapi translates the query and returns results. Below is an example query.

aliasByNode(exclude(exclude(exclude(nonNegativeDerivative(sumSeriesWithWildcards(scaleToSeconds(pcs.prod-decaf-us-west-2.pcs-*.route.*.GET.status.*.count, 1), 2, 4)), '2XX'), '4XX'), '5XX'), 5)

pcs-* translates to pod directories that are over 5000 in number and have timestamps for previous 30 days (retention). 

Just wanted to check if a time filter of 24 hours would scan all of these directories and return respective data or it just ignores directories that have timestamp older than 24 hours and scans only directories within time filter.

**Comment by Felixoid** [comment]
Hey. Carbonapi doesn't do anything like that, it sends the request to backends. And different backends use different logic.

- Whisper always scans all directories.
- graphite-clickhouse has settings to query daily index or to behave like whisper and query the global index

Those are to I'm aware of.

**Reply by rickyari** [reply]
Does this mean we are using whisper
backendsv2:
        backends:
          -
            groupName: "gocarbon"
            protocol: "carbonapi_v3_pb"
            lbMethod: "broadcast"
            maxTries: 3
            maxBatchSize: 100
            servers:


And also to mention we are using go-carbon as backend.

**Reply by Felixoid** [reply]
you use go-carbon :-j

And yes, it works on top of whisper. First search for every matching metric (`dir.dir.dir.file`) file (`dir/dir/dir/file.wsp`), and only then queries the date range for each one

**Reply by rickyari** [reply]
Thanks for the information. Is it possible to optimize this behaviour by keeping the same backend go-carbon

**Reply by Felixoid** [reply]
I am quite biased on using ClickHouse based backends, and would call myself an expert in it.

For go-carbon optimization it's best to ask @Civil

**Reply by deniszh** [reply]
> Just wanted to check if a time filter of 24 hours would scan all of these directories and return respective data or it just ignores directories that have timestamp older than 24 hours and scans only directories within time filter.

I'm afraid optimization above not really possible for whisper. It's not storing any time information in path. But you can try to enable different indices and optimizations for go-carbon, it can help. Please check read me for go-carbon config - https://github.com/go-graphite/go-carbon/blob/master/deploy/go-carbon.conf
